lib/protocol/protocol.js

```diff
@@ -259,12 +262,29 @@ function buildHandler (app, method, path, spec, driver, isSessCmd) {
         driverRes = await driver.execute(spec.command, ...args);
       }
 
+      // Get the protocol after executeCommand (when command is `createSession`, protocol is assigned within
+      // createSession function)
+      isW3C = driver.isW3CProtocol();
+      isMJSONWP = driver.isMjsonwpProtocol();
+
+      // If `executeCommand` was overridden and the method returns an object
+      // with a protocol and value/error property, re-assign the protocol
+      if (_.isObject(driverRes) && _.isString(driverRes.protocol) && (!_.isNil(driverRes.value) || _.isError(driverRes.error))) {
```

Comment:
From a software engineering standpoint, I'm not a big fan of this class knowing all this info about `BaseDriver`. I'd rather `BaseDriver#executeCommand` add the appropriate info to its result, so we can just check the protocol from it.

Reply:
That's a good idea. You're thinking that `executeCommand` should _always_ return an object of the form:

```javascript
{
  protocol: <W3C | MJSONWP>,
  value: <response value>,
  error: <response error>,
}
```

That would be do-able because afaik `appium/appium` is the only method that extends `executeCommand` and thus we wouldn't break anything else.

Reply:
I see usages of `executeCommand` in base-driver, appium itself, ios-driver, and xcuitest-driver

Reply:
Yeah... what if we made a new method called something like `executeCommandWithProtocol` that returns the protocol and value and base driver looks for that method first 

I don't think there is a really clean solution that doesn't involve heavy rewriting.

Reply:
Adding another method that would be called in preference would not work, since the subclasses that extend `executeCommand` still need that method to be called (in ios and xcuitest cases, for async script operation).